sdks/python/apache_beam/transforms/external.py

```diff
@@ -973,6 +980,68 @@ def __init__(
         path_to_jar, extra_args, classpath=classpath, append_args=append_args)
 
 
+def _may_be_use_transform_service(provided_service=None, options=None):
+  # For anything other than 'JavaJarExpansionService' we just use the
+  # provided service. For example, string address of an already available
+  # service.
+  if not isinstance(provided_service, JavaJarExpansionService):
+    return provided_service
+
+  def is_java_available():
+    cmd = ['javaxx', '--version']
+
+    try:
+      subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    except:  # pylint: disable=bare-except
+      return False
+
+    return True
+
+  def is_docker_available():
+    cmd = ['docker', '--version']
+
+    try:
+      subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    except:  # pylint: disable=bare-except
+      return False
+
+    return True
+
+  # We try java and docker based expansion services in that order.
+
+  java_available = is_java_available()
+  docker_available = is_docker_available()
+
+  use_transform_service = options.view_as(
+      CrossLanguageOptions).use_transform_service
+
+  if (java_available and provided_service
+      and not use_transform_service):
+    return provided_service
+  elif docker_available:
+    if use_transform_service:
+      error_append = 'it was explicitly requested'
+    elif not java_available:
+      error_append = 'java is not available in the system'
+    else:
+      error_append = 'a Java expansion service was not provided.'
+
+    logging.info(
+        'Trying to expand the external transform using the Docker-based '
+        'transform service since %s.' % error_append)
+    project_name = str(uuid.uuid4())
+    port = subprocess_server.pick_port(None)[0]
```

Comment:
Do you mind adding project_name and port to the log? i feel it might help with debugging